<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data Science Workshop</title>
  <meta name="description" content="Lessons for the BSPR 2018 Data Science Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Data Science Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data Science Workshop" />
  
  <meta name="twitter:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

<meta name="author" content="Alistair Bailey">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="import.html">
<link rel="next" href="going-further.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#requirements"><i class="fa fa-check"></i>Requirements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#what-are-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> What are R and RStudio?</a><ul>
<li class="chapter" data-level="1.1.1" data-path="intro.html"><a href="intro.html#environments"><i class="fa fa-check"></i><b>1.1.1</b> Environments</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#why-learn-r-or-any-language"><i class="fa fa-check"></i><b>1.2</b> Why learn R, or any language ?</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#finding-your-way-around-rstudio"><i class="fa fa-check"></i><b>1.3</b> Finding your way around RStudio</a><ul>
<li class="chapter" data-level="1.3.1" data-path="intro.html"><a href="intro.html#what-is-real"><i class="fa fa-check"></i><b>1.3.1</b> What is real?</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="intro.html"><a href="intro.html#where-am-i"><i class="fa fa-check"></i><b>1.4</b> Where am I?</a></li>
<li class="chapter" data-level="1.5" data-path="intro.html"><a href="intro.html#r-projects"><i class="fa fa-check"></i><b>1.5</b> R projects</a></li>
<li class="chapter" data-level="1.6" data-path="intro.html"><a href="intro.html#names"><i class="fa fa-check"></i><b>1.6</b> Naming things</a></li>
<li class="chapter" data-level="1.7" data-path="intro.html"><a href="intro.html#seeking-help"><i class="fa fa-check"></i><b>1.7</b> Seeking help</a><ul>
<li class="chapter" data-level="1.7.1" data-path="intro.html"><a href="intro.html#asking-for-help"><i class="fa fa-check"></i><b>1.7.1</b> Asking for help</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="tidyverse.html"><a href="tidyverse.html"><i class="fa fa-check"></i><b>2</b> Getting started in R and the tidyverse</a><ul>
<li class="chapter" data-level="2.1" data-path="tidyverse.html"><a href="tidyverse.html#the-tidyverse-and-tidy-data"><i class="fa fa-check"></i><b>2.1</b> The tidyverse and tidy data</a></li>
<li class="chapter" data-level="2.2" data-path="tidyverse.html"><a href="tidyverse.html#data-visualisation"><i class="fa fa-check"></i><b>2.2</b> Data visualisation</a></li>
<li class="chapter" data-level="2.3" data-path="tidyverse.html"><a href="tidyverse.html#workflow-basics"><i class="fa fa-check"></i><b>2.3</b> Workflow basics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="tidyverse.html"><a href="tidyverse.html#assigning-objects"><i class="fa fa-check"></i><b>2.3.1</b> Assigning objects</a></li>
<li class="chapter" data-level="2.3.2" data-path="tidyverse.html"><a href="tidyverse.html#function-anatomy"><i class="fa fa-check"></i><b>2.3.2</b> Function anatomy</a></li>
<li class="chapter" data-level="2.3.3" data-path="tidyverse.html"><a href="tidyverse.html#atomic-vectors"><i class="fa fa-check"></i><b>2.3.3</b> Atomic vectors</a></li>
<li class="chapter" data-level="2.3.4" data-path="tidyverse.html"><a href="tidyverse.html#attributes"><i class="fa fa-check"></i><b>2.3.4</b> Attributes</a></li>
<li class="chapter" data-level="2.3.5" data-path="tidyverse.html"><a href="tidyverse.html#factors"><i class="fa fa-check"></i><b>2.3.5</b> Factors</a></li>
<li class="chapter" data-level="2.3.6" data-path="tidyverse.html"><a href="tidyverse.html#lists"><i class="fa fa-check"></i><b>2.3.6</b> Lists</a></li>
<li class="chapter" data-level="2.3.7" data-path="tidyverse.html"><a href="tidyverse.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.3.7</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.3.8" data-path="tidyverse.html"><a href="tidyverse.html#data-frames"><i class="fa fa-check"></i><b>2.3.8</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="tidyverse.html"><a href="tidyverse.html#learning-more-r"><i class="fa fa-check"></i><b>2.4</b> Learning more R</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="import.html"><a href="import.html"><i class="fa fa-check"></i><b>3</b> Creating scripts and importing data</a><ul>
<li class="chapter" data-level="3.1" data-path="import.html"><a href="import.html#some-definitions"><i class="fa fa-check"></i><b>3.1</b> Some definitions</a></li>
<li class="chapter" data-level="3.2" data-path="import.html"><a href="import.html#using-scripts"><i class="fa fa-check"></i><b>3.2</b> Using scripts</a></li>
<li class="chapter" data-level="3.3" data-path="import.html"><a href="import.html#running-code"><i class="fa fa-check"></i><b>3.3</b> Running code</a></li>
<li class="chapter" data-level="3.4" data-path="import.html"><a href="import.html#creating-a-r-script"><i class="fa fa-check"></i><b>3.4</b> Creating a R script</a></li>
<li class="chapter" data-level="3.5" data-path="import.html"><a href="import.html#setting-up-our-environment"><i class="fa fa-check"></i><b>3.5</b> Setting up our environment</a><ul>
<li class="chapter" data-level="3.5.1" data-path="import.html"><a href="import.html#bioconductor"><i class="fa fa-check"></i><b>3.5.1</b> Bioconductor</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="import.html"><a href="import.html#importing-data"><i class="fa fa-check"></i><b>3.6</b> Importing data</a></li>
<li class="chapter" data-level="3.7" data-path="import.html"><a href="import.html#exploring-the-data"><i class="fa fa-check"></i><b>3.7</b> Exploring the data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="transform.html"><a href="transform.html"><i class="fa fa-check"></i><b>4</b> Transformation and visualisation</a><ul>
<li class="chapter" data-level="4.1" data-path="transform.html"><a href="transform.html#fold-change-and-log-fold-change"><i class="fa fa-check"></i><b>4.1</b> Fold change and log-fold change</a></li>
<li class="chapter" data-level="4.2" data-path="transform.html"><a href="transform.html#missing-values"><i class="fa fa-check"></i><b>4.2</b> Dealing with missing values</a></li>
<li class="chapter" data-level="4.3" data-path="transform.html"><a href="transform.html#normalisation"><i class="fa fa-check"></i><b>4.3</b> Data normalization</a><ul>
<li class="chapter" data-level="4.3.1" data-path="transform.html"><a href="transform.html#t-test"><i class="fa fa-check"></i><b>4.3.1</b> T-test</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="transform.html"><a href="transform.html#visualising-data"><i class="fa fa-check"></i><b>4.4</b> Visualising data</a></li>
<li class="chapter" data-level="4.5" data-path="transform.html"><a href="transform.html#creating-a-volcano-plot"><i class="fa fa-check"></i><b>4.5</b> Creating a volcano plot</a></li>
<li class="chapter" data-level="4.6" data-path="transform.html"><a href="transform.html#creating-a-heatmap-plot"><i class="fa fa-check"></i><b>4.6</b> Creating a heatmap plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>5</b> Going further</a><ul>
<li class="chapter" data-level="5.1" data-path="going-further.html"><a href="going-further.html#learning-dplyr-verbs"><i class="fa fa-check"></i><b>5.1</b> Learning <code>dplyr</code> verbs</a></li>
<li class="chapter" data-level="5.2" data-path="going-further.html"><a href="going-further.html#getting-help-and-joining-the-r-community"><i class="fa fa-check"></i><b>5.2</b> Getting help and joining the R community</a></li>
<li class="chapter" data-level="5.3" data-path="going-further.html"><a href="going-further.html#communication-creating-reports-presentations-and-websites"><i class="fa fa-check"></i><b>5.3</b> Communication: creating reports, presentations and websites</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data Science Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="transform" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Transformation and visualisation</h1>
<p>Having imported our data set of observations for 7702 proteins from cells in three control experiments and three treatment experiments. Remember, the observations are signal intensity measurements from the mass spectrometer, and these intensities relate to the amount of protein in each experiment and under each condition.</p>
<p>Next we will transform the data to examine the effect of the treatment on the cellular proteome and visualise the output using a volcano plot and a heatmap.</p>
<div id="fold-change-and-log-fold-change" class="section level2">
<h2><span class="header-section-number">4.1</span> Fold change and log-fold change</h2>
<p>Fold changes are ratios, the ratio of say protein expression before and after treatment, where a value larger than 1 for a protein implies that protein expression was greater after the treatment.</p>
<p>In life sciences, fold change is often reported as log-fold change. Why is that? There are at least two reasons which can be shown by plotting.</p>
<p>One is that ratios are not symmetrical around 1, so it’s difficult to observe both changes in the forwards and backwards direcion i.e. proteins where expression went up and proteins where expression went down due to treatment. When we transform ratios on a log scale, the scale becomes symmetric around 0 and thus we can now observe the distribution of ratios in terms of positive, negative or no change.</p>

<div class="figure"><span id="fig:fold-change-1"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-1-1.png" alt="Ratios are not symmetric around one, logratios are symmetric around zero." width="672" />
<p class="caption">
Figure 4.1: Ratios are not symmetric around one, logratios are symmetric around zero.
</p>
</div>
<p>A second reason is that transforming values onto a log scale changes where the numbers actually occur when plotted on that scale. If we consider the log scale to represent magnitudes, then we can more easily see changes of small and large magnitudes when we plot the data.</p>
<p>For example, a fold change of 32 times can be either a ratio 1/32 or 32/1.</p>
<p>As shown in Figure <a href="transform.html#fig:fold-change-2">4.2</a>, 1/32 is much closer to 1 than 32/1, but transformed to a log scale we see that in terms of magnitude of difference it is the same as 32/1.</p>

<div class="figure"><span id="fig:fold-change-2"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-2-1.png" alt="Transformation of scales using log transformation." width="672" />
<p class="caption">
Figure 4.2: Transformation of scales using log transformation.
</p>
</div>
</div>
<div id="missing-values" class="section level2">
<h2><span class="header-section-number">4.2</span> Dealing with missing values</h2>
<!---
1. Load the data, we need multiple replicates for each condition. To be tidy
each protein is a set of observations (rows) of the variables, which are the
values recorded for each replicate (columns).

2. Tidy up and deal with missing values, either impute or exclude missing values
and normalise. --->
<p>Unless we’re really lucky, it’s unlikely that we’ll get observations for the same numbers of proteins in all replicated experiments. This means there will be missing values for some proteins when looking at all the experiments together. This then raises the question of what to do about the missing values? We have two choices:</p>
<ol style="list-style-type: decimal">
<li>Only analyse the proteins that we have observations for in all experiments.</li>
<li>Impute values for the missing values from the existing observations.</li>
</ol>
<p>There are pros and cons to either approach. Here for simplicity we’ll use only the proteins for which we have observations in all assays.</p>
<p>We can drop the proteins with missing values by piping our data set to the <code>drop_na()</code> function from the <code>tidyr</code> package like so. We assign this to a new object called <code>dat_tidy</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Remove the missing values</span>
dat_tidy &lt;-<span class="st"> </span>dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">drop_na</span>()

<span class="co"># Nunber of proteins in original data</span>
dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">Number_of_proteins =</span> <span class="kw">n</span>())</code></pre></div>
<pre><code>## # A tibble: 1 x 1
##   Number_of_proteins
##                &lt;int&gt;
## 1               7702</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Nunber of proteins without missing values</span>
dat_tidy <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">Number_of_proteins =</span> <span class="kw">n</span>())</code></pre></div>
<pre><code>## # A tibble: 1 x 1
##   Number_of_proteins
##                &lt;int&gt;
## 1               1145</code></pre>
<p>This shrinks the dataset from 7,702 proteins to 1,145 proteins, so we can see why imputing the missing values might be more atrractive.</p>
<p>One approach you might like to try is to impute the data by replacing the missing values with the mean observation for each protein under each condition.</p>
</div>
<div id="normalisation" class="section level2">
<h2><span class="header-section-number">4.3</span> Data normalization</h2>
<p>To perform statistical inference, for example whether treatment increases or decreases protein abundance, we need to account for the variation that occurs from run to run on our spectrometers and each give rise to a different distribution. This is as opposed to variation arising from treatment versus control which we are interested in understanding. Hence normalization seeks to reduce the run-to-run sources of variation.</p>
<p>A method of normalization introduced for DNA microarray analysis is quantile normalization <span class="citation">(B. M. Bolstad et al. <a href="#ref-bolstad2003">2003</a>)</span>.</p>
<p>If we consider our proteomics data as a distribution of values, one value for the concentration of each protein in our experiment that together form a distribution. Figure <a href="transform.html#fig:data-dist">4.3</a> shows the distribution of protein concentrations observed for the three control and three treatment assays. As we can see the distributions are different for each assay.</p>

<div class="figure" style="text-align: center"><span id="fig:data-dist"></span>
<img src="bspr-workshop-2018_files/figure-html/data-dist-1.png" alt="Protein data for six assays plotted as a distributions." width="80%" />
<p class="caption">
Figure 4.3: Protein data for six assays plotted as a distributions.
</p>
</div>
<p>A quantile represents a region of distribution, for example the 0.95 quantile is the value such that 95% of the data lies below it. To normalize two or more distributions with each other without recourse to a reference distribution we:</p>
<ol style="list-style-type: lower-roman">
<li>Rank the value in each experiment (represented in the columns) from lowest to highest. In other words identify the quantiles for each protein in each experiment.</li>
<li>Sort each experiment (the columns) from lowest to highest value.</li>
<li>Calculate the mean across the rows for the sorted values.</li>
<li>Then substitute these mean values back according to rank for each experiment to restore the original order.</li>
</ol>
<p>This results in the highest ranking observation in each experiment becoming the mean of the highest observations across all experiments, the second ranking observation in each experiment becoming the mean of the second highest observations across all experiments. Therefore the distributions for each each experiment are now the same.</p>
<p><a href="https://davetang.org/muse/2014/07/07/quantile-normalisation-in-r/">Dave Tang’s Blog:Quantile Normalisation in R</a> has more details on this approach.</p>

<div class="figure" style="text-align: center"><span id="fig:quant-norm"></span>
<img src="img/quant_norm.png" alt="Quantile Normalisation from Rafael Irizarry’s tweet." width="80%" />
<p class="caption">
Figure 4.4: Quantile Normalisation from <a href="https://twitter.com/rafalab/status/545586012219772928?ref_src=twsrc%5Etfw">Rafael Irizarry’s tweet</a>.
</p>
</div>
<p>These result of quantile normalization is that our distributions become statisitcally identitical, which we can see by plotting the densities of the normalized data. As shown in Figure <a href="transform.html#fig:compare-normalisation">4.5</a> the distributions all overlay.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Quantile normalisation : the aim is to give different distributions the</span>
<span class="co"># same statistical properties</span>
quantile_normalisation &lt;-<span class="st"> </span><span class="cf">function</span>(df){
  df_rank &lt;-<span class="st"> </span><span class="kw">apply</span>(df,<span class="dv">2</span>,rank,<span class="dt">ties.method=</span><span class="st">&quot;average&quot;</span>)
  df_sorted &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">apply</span>(df, <span class="dv">2</span>, sort))
  df_mean &lt;-<span class="st"> </span><span class="kw">apply</span>(df_sorted, <span class="dv">1</span>, mean)
   
  index_to_mean &lt;-<span class="st"> </span><span class="cf">function</span>(my_index, my_mean){
    <span class="kw">return</span>(my_mean[my_index])
  }
   
  df_final &lt;-<span class="st"> </span><span class="kw">apply</span>(df_rank, <span class="dv">2</span>, index_to_mean, <span class="dt">my_mean=</span>df_mean) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">as.tibble</span>()
  
  <span class="kw">return</span>(df_final)
}</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_norm &lt;-<span class="st"> </span>dat_tidy <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">quantile_normalisation</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">bind_cols</span>(dat_tidy[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],.)</code></pre></div>

<div class="figure" style="text-align: center"><span id="fig:compare-normalisation"></span>
<img src="bspr-workshop-2018_files/figure-html/compare-normalisation-1.png" alt="Comparison of the protein distributions before normalization (left) and after quantile normalization (right)." width="80%" />
<p class="caption">
Figure 4.5: Comparison of the protein distributions before normalization (left) and after quantile normalization (right).
</p>
</div>
<div id="t-test" class="section level3">
<h3><span class="header-section-number">4.3.1</span> T-test</h3>
<p>Things we need to consider in performing our t-test:</p>
<ol style="list-style-type: decimal">
<li>We are perfoming a test for each protein for between three control and treatment experiments and assume unequal variances between the control and treatment for each protein. A look at a single protein supports this assumption:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Variance of first control protein observations</span>
<span class="kw">var</span>(<span class="kw">c</span>(dat_norm<span class="op">$</span>control_<span class="dv">1</span>[<span class="dv">1</span>], dat_norm<span class="op">$</span>control_<span class="dv">2</span>[<span class="dv">1</span>], dat_norm<span class="op">$</span>control_<span class="dv">3</span>[<span class="dv">1</span>]))</code></pre></div>
<pre><code>## [1] 0.01473403</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Variance of first treatment protein observations</span>
<span class="kw">var</span>(<span class="kw">c</span>(dat_norm<span class="op">$</span>treatment_<span class="dv">1</span>[<span class="dv">1</span>], dat_norm<span class="op">$</span>treatment_<span class="dv">2</span>[<span class="dv">1</span>], dat_norm<span class="op">$</span>treatment_<span class="dv">3</span>[<span class="dv">1</span>]))</code></pre></div>
<pre><code>## [1] 0.002110211</code></pre>
<p>Hence we will perform a Welch’s t-test for unequal variances.</p>
<ol start="2" style="list-style-type: decimal">
<li><p>We don’t know whether the effect of the treatment is to increase or decrease the concentration of the protein, hence we will perform a two-sided t-test.</p></li>
<li><p>The observations for the proteins are for proteins of the same type but from independent experiments, rather than observations of the same individuals before and after treatment. Hence we test the observations as unpaired samples.</p></li>
</ol>
<p>Use <code>t.test</code> to perform Welch Two Sample t-test on untransformed data. This outputs the p-values we need for each protein. I couldn’t figure out a tidy way to do this, so I am using the base R apply function here.</p>
<p>What’s going on in the code below is the following,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># T-test function for multiple experiments</span>
expriments_ttest &lt;-<span class="st"> </span><span class="cf">function</span>(dt,grp1,grp2){
  <span class="co"># Subset control group and convert to numeric</span>
  x &lt;-<span class="st"> </span>dt[grp1] <span class="op">%&gt;%</span><span class="st"> </span>unlist <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.numeric</span>()
  <span class="co"># Subset treatment group and convert to numeric</span>
  y &lt;-<span class="st"> </span>dt[grp2] <span class="op">%&gt;%</span><span class="st"> </span>unlist <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.numeric</span>()
  <span class="co"># Perform t-test using the mean of x and y</span>
  result &lt;-<span class="st"> </span><span class="kw">t.test</span>(x, y)
  <span class="co"># Return p-values</span>
  <span class="kw">return</span>(result<span class="op">$</span>p.value)
} </code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Apply t-test function to data</span>
<span class="co"># array = dat, 1 = rows, FUN = expriments_ttest, and arguements</span>
p_vals &lt;-<span class="st"> </span><span class="kw">apply</span>(dat_norm,<span class="dv">1</span>,expriments_ttest, <span class="dt">grp1 =</span> <span class="kw">c</span>(<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>), <span class="dt">grp2 =</span> <span class="kw">c</span>(<span class="dv">6</span><span class="op">:</span><span class="dv">8</span>))

<span class="co"># Calculate the p-value for the first protein and compare with function</span>
<span class="kw">paste0</span>(<span class="kw">round</span>(p_vals[<span class="dv">1</span>],<span class="dv">4</span>),<span class="st">&quot; &quot;</span>, 
       <span class="kw">round</span>(<span class="kw">t.test</span>(<span class="kw">as.numeric</span>(dat_norm[<span class="dv">1</span>,<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>]),
                    <span class="kw">as.numeric</span>(dat_norm[<span class="dv">1</span>,<span class="dv">6</span><span class="op">:</span><span class="dv">8</span>]))<span class="op">$</span>p.value,<span class="dv">4</span>))</code></pre></div>
<pre><code>## [1] &quot;0.0927 0.0927&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot histograms</span>
<span class="kw">hist</span>(p_vals)</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/t-tests-1.png" width="672" /></p>
<ol start="4" style="list-style-type: decimal">
<li>Perform log transformation of the observations for each protein.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Select columns and log data</span>
dat_log &lt;-<span class="st"> </span>dat_norm <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(protein_accession,protein_description)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">log2</span>()</code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>Calculate the mean observation for each protein under each condition.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">con &lt;-<span class="st"> </span><span class="kw">apply</span>(dat_log[,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],<span class="dv">1</span>,mean)
trt &lt;-<span class="st"> </span><span class="kw">apply</span>(dat_log[,<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>],<span class="dv">1</span>,mean)</code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>The log fold change is then the difference between condition 1 and condition 2.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot a histogram to look at the distribution.</span>
<span class="co"># Calculate fold change</span>
dat_fc &lt;-<span class="st"> </span>con <span class="op">-</span><span class="st"> </span>trt
<span class="co"># Plot histograms</span>
<span class="kw">hist</span>(dat_fc)</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/log-fc-1.png" width="672" /></p>
</div>
</div>
<div id="visualising-data" class="section level2">
<h2><span class="header-section-number">4.4</span> Visualising data</h2>
<p>Based on empircal research, there are some general rules on visulisations that are worth bearing in mind:</p>
<ol style="list-style-type: decimal">
<li>Plot</li>
</ol>
</div>
<div id="creating-a-volcano-plot" class="section level2">
<h2><span class="header-section-number">4.5</span> Creating a volcano plot</h2>
<!--- 7. Create a combined table of log fold change and p-values for all the proteins
for plotting a volcano plot. --->
<p>A volcano plot is a plot of the log fold change in the observation between two conditions on the x-axis, for example the protein expression between treatment and control conditions. On the y-axis is the corresponding p-value for each observation, representing the likelihood that an observed change is due to the different conditions rather than arising from a natural variation in the fold change that might be observed if we performed many replications of the experiment.</p>
<p>The aim of a volcano plot is to enable the viewer to quickly see the effect (if any) of an experiment with two conditions on many species (i.e. proteins) in terms of both an increase and decrease of the observed value.</p>
<p>Like all plots it has it’s good and bad points, namely it’s good that we can visualise a lot of complex information in one plot. However this is also it’s main weakness, it’s rather complicated to understand in one glance.</p>
<p>However, volcano plots are widely used in the literature, so there may be an amount of <a href="https://en.wikipedia.org/wiki/Social_proof">social proof</a> giving rise to their popularity as much as their utility.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_vplot &lt;-<span class="st"> </span><span class="kw">tibble</span>(<span class="dt">prots=</span> dat_norm<span class="op">$</span>protein_accession,
                   <span class="dt">logfc =</span> dat_fc,
                   <span class="dt">logpval =</span> <span class="op">-</span><span class="dv">1</span><span class="op">*</span><span class="kw">log10</span>(p_vals))

dat_vplot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(logfc,logpval)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/volcano-plot-1.png" width="672" /></p>
<p>However it would be much more useful with some extra formatting…</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_vplot <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Add a threhold for significant observations</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">threshold =</span> <span class="kw">if_else</span>(logfc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>logpval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span> <span class="op">|</span>
<span class="st">                               </span>logfc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>logpval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span>,<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Plot with points coloured according to the threshold</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(logfc,logpval, <span class="dt">colour =</span> threshold)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span><span class="co"># Alpha sets the transparency of the points</span>
<span class="st">  </span><span class="co"># Add dotted lines to indicate the threshold, semi-transparent</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">1.5</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">2</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="op">-</span><span class="dv">2</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set the colour of the points</span>
<span class="st">  </span><span class="kw">scale_colour_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>=<span class="st"> &quot;red&quot;</span>, <span class="st">&quot;B&quot;</span>=<span class="st"> &quot;black&quot;</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;log2 fold change&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;-log10 p-value&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="co"># Relabel the axes</span>
<span class="st">  </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span><span class="co"># Set the theme</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>) <span class="co"># Hide the legend</span></code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/nice-vplot-1.png" width="672" /></p>
<p>But which proteins are the significant observations?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_vplot <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Add a threhold for significant observations</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">threshold =</span> <span class="kw">if_else</span>(logfc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>logpval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span> <span class="op">|</span>
<span class="st">                               </span>logfc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>logpval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span>,<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>),
         <span class="dt">prot_id =</span> <span class="kw">str_extract</span>(prots,<span class="st">&quot;.{6}$&quot;</span>)) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># Get last six characters</span>
<span class="st">  </span><span class="co"># Filter observations above the threshold</span>
<span class="st">  </span><span class="kw">filter</span>(threshold <span class="op">==</span><span class="st"> &quot;A&quot;</span>)</code></pre></div>
<pre><code>## # A tibble: 5 x 5
##   prots               logfc logpval threshold prot_id
##   &lt;chr&gt;               &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;  
## 1 AKA12_HUMAN_Q02952  -2.29    1.52 A         Q02952 
## 2 GFPT2_HUMAN_O94808  -3.09    2.45 A         O94808 
## 3 H7BYV1_HUMAN_H7BYV1  2.05    2.46 A         H7BYV1 
## 4 ITAV_HUMAN_P06756   -2.22    2.62 A         P06756 
## 5 CHD5_HUMAN_Q8TDI0    5.04    1.60 A         Q8TDI0</code></pre>
</div>
<div id="creating-a-heatmap-plot" class="section level2">
<h2><span class="header-section-number">4.6</span> Creating a heatmap plot</h2>
<!-- 1. Clustering the data -->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Keep the same p-val cut-off, but relax the logfc to 1 which represents a </span>
<span class="co"># doubling</span>
dat_mut &lt;-<span class="st"> </span>dat_norm <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">logpval =</span> dat_vplot<span class="op">$</span>logpval, <span class="dt">logfc =</span> dat_vplot<span class="op">$</span>logfc) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(logpval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span> <span class="op">&amp;</span><span class="st"> </span>(logfc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">1</span> <span class="op">|</span><span class="st"> </span>logfc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">1</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">9</span><span class="op">:</span><span class="dv">10</span>))

dat_sel &lt;-<span class="st"> </span><span class="kw">as.matrix.data.frame</span>(dat_mut[,<span class="dv">2</span><span class="op">:</span><span class="dv">7</span>]) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">log2</span>()
<span class="kw">row.names</span>(dat_sel) &lt;-<span class="st"> </span>dat_mut<span class="op">$</span>protein_accession

dat.tn &lt;-<span class="st"> </span><span class="kw">scale</span>(<span class="kw">t</span>(dat_sel)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">t</span>()
<span class="co">#dat.tn &lt;- t(dat.n)</span>
<span class="co">#dat.tn &lt;- dat_sel</span>

<span class="co">#gplots::heatmap.2(dat.tn, scale = &#39;row&#39;,trace=&quot;none&quot;)</span>

<span class="kw">pheatmap</span>(dat.tn,<span class="dt">cutree_rows =</span> <span class="dv">2</span>,
         <span class="dt">cutree_cols =</span> <span class="dv">2</span>,
         <span class="dt">fontsize_row =</span> <span class="dv">6</span>)</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/heatmap-2-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cal_z_score &lt;-<span class="st"> </span><span class="cf">function</span>(x){
  (x <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(x)) <span class="op">/</span><span class="st"> </span><span class="kw">sd</span>(x)
}

data_subset_norm &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(dat_mut[,<span class="dv">2</span><span class="op">:</span><span class="dv">7</span>], <span class="dv">1</span>, cal_z_score))

data_subset_norm <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_colnames</span>(<span class="kw">c</span>(<span class="st">&quot;Ctl 1&quot;</span>, <span class="st">&quot;Ctl 2&quot;</span>, <span class="st">&quot;Ctl 3&quot;</span>,
                                    <span class="st">&quot;Trt 1&quot;</span>, <span class="st">&quot;Trt 2&quot;</span>, <span class="st">&quot;Trt 3&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_rownames</span>(dat_mut<span class="op">$</span>protein_accession) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">pheatmap</span>(.,
           <span class="dt">fontsize =</span> <span class="dv">6</span>,
           <span class="dt">cutree_rows =</span> <span class="dv">2</span>,
           <span class="dt">cutree_cols =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/heatmap-2-2.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d1 &lt;-<span class="st"> </span>dat.tn <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">t</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">dist</span>(.,<span class="dt">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">diag =</span> <span class="ot">FALSE</span>, <span class="dt">upper =</span> <span class="ot">FALSE</span>)

d2 &lt;-<span class="st"> </span>dat.tn <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">dist</span>(.,<span class="dt">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">diag =</span> <span class="ot">FALSE</span>, <span class="dt">upper =</span> <span class="ot">FALSE</span>)

<span class="co"># Clustering distance between experiments using Ward linkage</span>
c1 &lt;-<span class="st"> </span><span class="kw">hclust</span>(d1, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="dt">members =</span> <span class="ot">NULL</span>)
<span class="co"># Clustering distance between proteins using Ward linkage</span>
c2 &lt;-<span class="st"> </span><span class="kw">hclust</span>(d2, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="dt">members =</span> <span class="ot">NULL</span>)

<span class="co"># Check clustering by plotting dendrograms</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">cex=</span><span class="fl">0.5</span>) <span class="co"># Make 2 rows, 1 col plot frame and shrink labels</span>
<span class="kw">plot</span>(c1); <span class="kw">plot</span>(c2) <span class="co"># Plot both cluster dendrograms</span></code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/heatmap-2-3.png" width="672" /></p>
<!-- 2. Plot the data -->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Set colours for heatmap, 25 increments</span>
my_palette &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;white&quot;</span>,<span class="st">&quot;red&quot;</span>))(<span class="dt">n =</span> <span class="dv">25</span>)

<span class="co"># Plot heatmap with heatmap.2</span>
<span class="kw">par</span>(<span class="dt">cex.main=</span><span class="fl">0.75</span>) <span class="co"># Shrink title fonts on plot</span>
dat.tn <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_colnames</span>(<span class="kw">c</span>(<span class="st">&quot;Ctl 1&quot;</span>, <span class="st">&quot;Ctl 2&quot;</span>, <span class="st">&quot;Ctl 3&quot;</span>,
                                    <span class="st">&quot;Trt 1&quot;</span>, <span class="st">&quot;Trt 2&quot;</span>, <span class="st">&quot;Trt 3&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_rownames</span>(dat_mut<span class="op">$</span>protein_accession) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>gplots<span class="op">::</span><span class="kw">heatmap.2</span>(.,                     <span class="co"># Tidy, normalised data</span>
          <span class="dt">Colv=</span><span class="kw">as.dendrogram</span>(c1),     <span class="co"># Experiments clusters in cols</span>
          <span class="dt">Rowv=</span><span class="kw">as.dendrogram</span>(c2),     <span class="co"># Protein clusters in rows</span>
          <span class="dt">density.info=</span><span class="st">&quot;histogram&quot;</span>,   <span class="co"># Plot histogram of data and colour key</span>
          <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>,               <span class="co"># Turn of trace lines from heat map</span>
          <span class="dt">col =</span> my_palette,           <span class="co"># Use my colour scheme</span>
          <span class="dt">cexRow=</span><span class="fl">0.3</span>,<span class="dt">cexCol=</span><span class="fl">0.75</span>)     <span class="co"># Amend row and column label fonts</span></code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/heatmap-1-1.png" width="672" /></p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-bolstad2003">
<p>Bolstad, B M, R A Irizarry, M Astrand, and T P Speed. 2003. “A Comparison of Normalization Methods for High Density Oligonucleotide Array Data Based on Variance and Bias.” <em>Bioinformatics (Oxford, England)</em> 19 (2): 185–93.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="import.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="going-further.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["bspr-workshop-2018.pdf", "bspr-workshop-2018.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
