<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Data Science Workshop</title>
  <meta name="description" content="Lessons for the BSPR 2018 Data Science Workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Data Science Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Data Science Workshop" />
  
  <meta name="twitter:description" content="Lessons for the BSPR 2018 Data Science Workshop" />
  

<meta name="author" content="Alistair Bailey">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="dplyr.html">
<link rel="next" href="going-further.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#requirements"><i class="fa fa-check"></i>Requirements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="intro.html"><a href="intro.html#r-rstudio"><i class="fa fa-check"></i><b>1.1</b> What are R and RStudio?</a><ul>
<li class="chapter" data-level="1.1.1" data-path="intro.html"><a href="intro.html#environments"><i class="fa fa-check"></i><b>1.1.1</b> Environments</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="intro.html"><a href="intro.html#why-learn-r-or-any-language"><i class="fa fa-check"></i><b>1.2</b> Why learn R, or any language ?</a></li>
<li class="chapter" data-level="1.3" data-path="intro.html"><a href="intro.html#finding-your-way-around-rstudio"><i class="fa fa-check"></i><b>1.3</b> Finding your way around RStudio</a><ul>
<li class="chapter" data-level="1.3.1" data-path="intro.html"><a href="intro.html#what-is-real"><i class="fa fa-check"></i><b>1.3.1</b> What is real?</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="intro.html"><a href="intro.html#where-am-i"><i class="fa fa-check"></i><b>1.4</b> Where am I?</a></li>
<li class="chapter" data-level="1.5" data-path="intro.html"><a href="intro.html#r-projects"><i class="fa fa-check"></i><b>1.5</b> R projects</a></li>
<li class="chapter" data-level="1.6" data-path="intro.html"><a href="intro.html#names"><i class="fa fa-check"></i><b>1.6</b> Naming things</a></li>
<li class="chapter" data-level="1.7" data-path="intro.html"><a href="intro.html#seeking-help"><i class="fa fa-check"></i><b>1.7</b> Seeking help</a><ul>
<li class="chapter" data-level="1.7.1" data-path="intro.html"><a href="intro.html#asking-for-help"><i class="fa fa-check"></i><b>1.7.1</b> Asking for help</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="tidyverse.html"><a href="tidyverse.html"><i class="fa fa-check"></i><b>2</b> Getting started in R and the tidyverse</a><ul>
<li class="chapter" data-level="2.1" data-path="tidyverse.html"><a href="tidyverse.html#the-tidyverse-and-tidy-data"><i class="fa fa-check"></i><b>2.1</b> The tidyverse and tidy data</a></li>
<li class="chapter" data-level="2.2" data-path="tidyverse.html"><a href="tidyverse.html#data-visualisation"><i class="fa fa-check"></i><b>2.2</b> Data visualisation</a><ul>
<li class="chapter" data-level="2.2.1" data-path="tidyverse.html"><a href="tidyverse.html#visualisations"><i class="fa fa-check"></i><b>2.2.1</b> Visualisations</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="tidyverse.html"><a href="tidyverse.html#workflow-basics"><i class="fa fa-check"></i><b>2.3</b> Workflow basics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="tidyverse.html"><a href="tidyverse.html#assigning-objects"><i class="fa fa-check"></i><b>2.3.1</b> Assigning objects</a></li>
<li class="chapter" data-level="2.3.2" data-path="tidyverse.html"><a href="tidyverse.html#function-anatomy"><i class="fa fa-check"></i><b>2.3.2</b> Function anatomy</a></li>
<li class="chapter" data-level="2.3.3" data-path="tidyverse.html"><a href="tidyverse.html#atomics"><i class="fa fa-check"></i><b>2.3.3</b> Atomic vectors</a></li>
<li class="chapter" data-level="2.3.4" data-path="tidyverse.html"><a href="tidyverse.html#attributes"><i class="fa fa-check"></i><b>2.3.4</b> Attributes</a></li>
<li class="chapter" data-level="2.3.5" data-path="tidyverse.html"><a href="tidyverse.html#factors"><i class="fa fa-check"></i><b>2.3.5</b> Factors</a></li>
<li class="chapter" data-level="2.3.6" data-path="tidyverse.html"><a href="tidyverse.html#lists"><i class="fa fa-check"></i><b>2.3.6</b> Lists</a></li>
<li class="chapter" data-level="2.3.7" data-path="tidyverse.html"><a href="tidyverse.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.3.7</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.3.8" data-path="tidyverse.html"><a href="tidyverse.html#data-frames"><i class="fa fa-check"></i><b>2.3.8</b> Data frames</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="tidyverse.html"><a href="tidyverse.html#learning-more-r"><i class="fa fa-check"></i><b>2.4</b> Learning more R</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="import.html"><a href="import.html"><i class="fa fa-check"></i><b>3</b> Creating scripts and importing data</a><ul>
<li class="chapter" data-level="3.1" data-path="import.html"><a href="import.html#some-definitions"><i class="fa fa-check"></i><b>3.1</b> Some definitions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="import.html"><a href="import.html#file-formats"><i class="fa fa-check"></i><b>3.1.1</b> Rectangular data and flat formats</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="import.html"><a href="import.html#using-scripts"><i class="fa fa-check"></i><b>3.2</b> Using scripts</a></li>
<li class="chapter" data-level="3.3" data-path="import.html"><a href="import.html#running-code"><i class="fa fa-check"></i><b>3.3</b> Running code</a></li>
<li class="chapter" data-level="3.4" data-path="import.html"><a href="import.html#creating-a-r-script"><i class="fa fa-check"></i><b>3.4</b> Creating a R script</a></li>
<li class="chapter" data-level="3.5" data-path="import.html"><a href="import.html#setting-up-our-environment"><i class="fa fa-check"></i><b>3.5</b> Setting up our environment</a><ul>
<li class="chapter" data-level="3.5.1" data-path="import.html"><a href="import.html#biocondutor"><i class="fa fa-check"></i><b>3.5.1</b> Bioconductor</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="import.html"><a href="import.html#importing-data"><i class="fa fa-check"></i><b>3.6</b> Importing data</a></li>
<li class="chapter" data-level="3.7" data-path="import.html"><a href="import.html#exploring-the-data"><i class="fa fa-check"></i><b>3.7</b> Exploring the data</a><ul>
<li class="chapter" data-level="3.7.1" data-path="import.html"><a href="import.html#glimpse-head-and-str"><i class="fa fa-check"></i><b>3.7.1</b> <code>glimpse</code>, <code>head</code> and <code>str</code></a></li>
<li class="chapter" data-level="3.7.2" data-path="import.html"><a href="import.html#summary-statisitics"><i class="fa fa-check"></i><b>3.7.2</b> Summary statisitics</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="dplyr.html"><a href="dplyr.html"><i class="fa fa-check"></i><b>4</b> <code>dplyr</code> verbs and piping</a><ul>
<li class="chapter" data-level="4.1" data-path="dplyr.html"><a href="dplyr.html#pipes"><i class="fa fa-check"></i><b>4.1</b> Pipes</a></li>
<li class="chapter" data-level="4.2" data-path="dplyr.html"><a href="dplyr.html#filter"><i class="fa fa-check"></i><b>4.2</b> Filter rows</a></li>
<li class="chapter" data-level="4.3" data-path="dplyr.html"><a href="dplyr.html#arrange-rows"><i class="fa fa-check"></i><b>4.3</b> Arrange rows</a></li>
<li class="chapter" data-level="4.4" data-path="dplyr.html"><a href="dplyr.html#select-columns"><i class="fa fa-check"></i><b>4.4</b> Select columns</a></li>
<li class="chapter" data-level="4.5" data-path="dplyr.html"><a href="dplyr.html#mutate"><i class="fa fa-check"></i><b>4.5</b> Create new variables</a></li>
<li class="chapter" data-level="4.6" data-path="dplyr.html"><a href="dplyr.html#create-grouped-summaries"><i class="fa fa-check"></i><b>4.6</b> Create grouped summaries</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="transform.html"><a href="transform.html"><i class="fa fa-check"></i><b>5</b> Transforming and visualising proteomics data</a><ul>
<li class="chapter" data-level="5.1" data-path="transform.html"><a href="transform.html#fold-change-and-log-fold-change"><i class="fa fa-check"></i><b>5.1</b> Fold change and log-fold change</a></li>
<li class="chapter" data-level="5.2" data-path="transform.html"><a href="transform.html#missing-values"><i class="fa fa-check"></i><b>5.2</b> Dealing with missing values</a></li>
<li class="chapter" data-level="5.3" data-path="transform.html"><a href="transform.html#normalisation"><i class="fa fa-check"></i><b>5.3</b> Data normalization</a></li>
<li class="chapter" data-level="5.4" data-path="transform.html"><a href="transform.html#hypothesis-testing-with-the-t-test"><i class="fa fa-check"></i><b>5.4</b> Hypothesis testing with the t-test</a></li>
<li class="chapter" data-level="5.5" data-path="transform.html"><a href="transform.html#calculating-fold-change"><i class="fa fa-check"></i><b>5.5</b> Calculating fold change</a></li>
<li class="chapter" data-level="5.6" data-path="transform.html"><a href="transform.html#visualising-the-transformed-data"><i class="fa fa-check"></i><b>5.6</b> Visualising the transformed data</a></li>
<li class="chapter" data-level="5.7" data-path="transform.html"><a href="transform.html#volcano-plot"><i class="fa fa-check"></i><b>5.7</b> Volcano plot</a><ul>
<li class="chapter" data-level="5.7.1" data-path="transform.html"><a href="transform.html#but-which-proteins-are-the-significant-observations"><i class="fa fa-check"></i><b>5.7.1</b> But which proteins are the significant observations?</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="transform.html"><a href="transform.html#creating-a-heatmap"><i class="fa fa-check"></i><b>5.8</b> Creating a heatmap</a><ul>
<li class="chapter" data-level="5.8.1" data-path="transform.html"><a href="transform.html#calculating-similarity-and-clustering"><i class="fa fa-check"></i><b>5.8.1</b> Calculating similarity and clustering</a></li>
<li class="chapter" data-level="5.8.2" data-path="transform.html"><a href="transform.html#plotting-the-heatmap"><i class="fa fa-check"></i><b>5.8.2</b> Plotting the heatmap</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="transform.html"><a href="transform.html#venn-diagram"><i class="fa fa-check"></i><b>5.9</b> Venn diagram</a></li>
<li class="chapter" data-level="5.10" data-path="transform.html"><a href="transform.html#peptide-sequence-logos"><i class="fa fa-check"></i><b>5.10</b> Peptide sequence logos</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="going-further.html"><a href="going-further.html"><i class="fa fa-check"></i><b>6</b> Going further</a><ul>
<li class="chapter" data-level="6.1" data-path="going-further.html"><a href="going-further.html#exporting-figures"><i class="fa fa-check"></i><b>6.1</b> Exporting figures</a></li>
<li class="chapter" data-level="6.2" data-path="going-further.html"><a href="going-further.html#exporting-data"><i class="fa fa-check"></i><b>6.2</b> Exporting data</a><ul>
<li class="chapter" data-level="6.2.1" data-path="going-further.html"><a href="going-further.html#writing-to-a-file"><i class="fa fa-check"></i><b>6.2.1</b> Writing to a file</a></li>
<li class="chapter" data-level="6.2.2" data-path="going-further.html"><a href="going-further.html#for-r"><i class="fa fa-check"></i><b>6.2.2</b> For R</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="going-further.html"><a href="going-further.html#joining-the-r-community"><i class="fa fa-check"></i><b>6.3</b> Joining the R community</a></li>
<li class="chapter" data-level="6.4" data-path="going-further.html"><a href="going-further.html#communication-creating-reports-presentations-and-websites"><i class="fa fa-check"></i><b>6.4</b> Communication: creating reports, presentations and websites</a><ul>
<li class="chapter" data-level="6.4.1" data-path="going-further.html"><a href="going-further.html#using-bookdown-to-write-a-thesis-dissertaion"><i class="fa fa-check"></i><b>6.4.1</b> Using bookdown to write a thesis dissertaion</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="going-further.html"><a href="going-further.html#machine-learning"><i class="fa fa-check"></i><b>6.5</b> Machine Learning</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Data Science Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="transform" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Transforming and visualising proteomics data</h1>
<p>Having imported our data set of observations for 7702 proteins from cells in three control experiments and three treatment experiments. Remember, the observations are signal intensity measurements from the mass spectrometer, and these intensities relate to the amount of protein in each experiment and under each condition.</p>
<p>Now we will transform the data to examine the effect of the treatment on the cellular proteome and visualise the output using a volcano plot and a heatmap. The hypothesis we are testing is that treatment changes the concentration of protein we observe.</p>
<p>A volcano plot is commonly used way of plotting changes in observed values on the x-axis against the likelihood of observing that change due to chance on the y-axis. Heatmaps are another way of visualising the relative (increase and decrease of) amounts of observed values.</p>
<div id="fold-change-and-log-fold-change" class="section level2">
<h2><span class="header-section-number">5.1</span> Fold change and log-fold change</h2>
<p>Fold changes are ratios, the ratio of say protein expression before and after treatment, where a value larger than 1 for a protein implies that protein expression was greater after the treatment.</p>
<p>In life sciences, fold change is often reported as log-fold change. Why is that? There are at least two reasons which can be shown by plotting.</p>
<p>One is that ratios are not symmetrical around 1, so it’s difficult to observe both changes in the forwards and backwards direcion i.e. proteins where expression went up and proteins where expression went down due to treatment. When we transform ratios on a log scale, the scale becomes symmetric around 0 and thus we can now observe the distribution of ratios in terms of positive, negative or no change.</p>

<div class="figure"><span id="fig:fold-change-1"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-1-1.png" alt="Ratios are not symmetric around one, logratios are symmetric around zero." width="672" />
<p class="caption">
Figure 5.1: Ratios are not symmetric around one, logratios are symmetric around zero.
</p>
</div>
<p>A second reason is that transforming values onto a log scale changes where the numbers actually occur when plotted on that scale. If we consider the log scale to represent magnitudes, then we can more easily see changes of small and large magnitudes when we plot the data.</p>
<p>For example, a fold change of 32 times can be either a ratio 1/32 or 32/1.</p>
<p>As shown in Figure <a href="transform.html#fig:fold-change-2">5.2</a>, 1/32 is much closer to 1 than 32/1, but transformed to a log scale we see that in terms of magnitude of difference it is the same as 32/1.</p>
<p>Often the log transformation is to a base of 2 as each increment of 1 represents a doubling, but sometimes a base of 10 is used, for example for p-values.</p>

<div class="figure"><span id="fig:fold-change-2"></span>
<img src="bspr-workshop-2018_files/figure-html/fold-change-2-1.png" alt="Transformation of scales using log transformation." width="672" />
<p class="caption">
Figure 5.2: Transformation of scales using log transformation.
</p>
</div>
</div>
<div id="missing-values" class="section level2">
<h2><span class="header-section-number">5.2</span> Dealing with missing values</h2>
<!---
1. Load the data, we need multiple replicates for each condition. To be tidy
each protein is a set of observations (rows) of the variables, which are the
values recorded for each replicate (columns).

2. Tidy up and deal with missing values, either impute or exclude missing values
and normalise. --->
<p>Unless we’re really lucky, it’s unlikely that we’ll get observations for the same numbers of proteins in all replicated experiments. This means there will be missing values for some proteins when looking at all the experiments together. This then raises the question of what to do about the missing values? We have two choices:</p>
<ol style="list-style-type: decimal">
<li>Only analyse the proteins that we have observations for in all experiments.</li>
<li>Impute values for the missing values from the existing observations.</li>
</ol>
<p>There are pros and cons to either approach. Here for simplicity we’ll use only the proteins for which we have observations in all assays.</p>
<p>We can drop the proteins with missing values by piping our data set to the <code>drop_na()</code> function from the <code>tidyr</code> package like so. We assign this to a new object called <code>dat_tidy</code>.</p>
<p>We’ll use the summarise function to compare the number of proteins before and after dropping the missing values using the <code>n()</code> counting function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Remove the missing values</span>
dat_tidy &lt;-<span class="st"> </span>dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">drop_na</span>()
<span class="co"># Nunber of proteins in original data</span>
dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">Number_of_proteins =</span> <span class="kw">n</span>())</code></pre></div>
<pre><code>## # A tibble: 1 x 1
##   Number_of_proteins
##                &lt;int&gt;
## 1               7702</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Nunber of proteins without missing values</span>
dat_tidy <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">Number_of_proteins =</span> <span class="kw">n</span>())</code></pre></div>
<pre><code>## # A tibble: 1 x 1
##   Number_of_proteins
##                &lt;int&gt;
## 1               1145</code></pre>
<p>This shrinks the dataset from 7,702 proteins to 1,145 proteins, so we can see why imputing the missing values might be more atrractive.</p>
<p>One approach you might like to try is to impute the data by replacing the missing values with the mean observation for each protein under each condition.</p>
</div>
<div id="normalisation" class="section level2">
<h2><span class="header-section-number">5.3</span> Data normalization</h2>
<p>To perform statistical inference, for example whether treatment increases or decreases protein abundance, we need to account for the variation that occurs from run to run on our spectrometers and each give rise to a different distribution. This is as opposed to variation arising from treatment versus control which we are interested in understanding. Hence normalisation seeks to reduce the run-to-run sources of variation.</p>
<p>A method of normalization introduced for DNA microarray analysis is quantile normalisation <span class="citation">(B. M. Bolstad et al. <a href="#ref-bolstad2003">2003</a>)</span>. There are various ways to normalise data, so using quantile normalisation here is primarily to demonstate the approach in R, you should consider what is best for your data.</p>
<p>If we consider our proteomics data as a distribution of values, one value for the concentration of each protein in our experiment that together form a distribution. Figure <a href="transform.html#fig:data-dist">5.3</a> shows the distribution of protein concentrations observed for the three control and three treatment assays. As we can see the distributions are different for each assay.</p>

<div class="figure" style="text-align: center"><span id="fig:data-dist"></span>
<img src="bspr-workshop-2018_files/figure-html/data-dist-1.png" alt="Protein data for six assays plotted as a distributions." width="80%" />
<p class="caption">
Figure 5.3: Protein data for six assays plotted as a distributions.
</p>
</div>
<p>A quantile represents a region of distribution, for example the 0.95 quantile is the value such that 95% of the data lies below it. To normalize two or more distributions with each other without recourse to a reference distribution we:</p>
<ol style="list-style-type: lower-roman">
<li>Rank the value in each experiment (represented in the columns) from lowest to highest. In other words identify the quantiles for each protein in each experiment.</li>
<li>Sort each experiment (the columns) from lowest to highest value.</li>
<li>Calculate the mean across the rows for the sorted values.</li>
<li>Then substitute these mean values back according to rank for each experiment to restore the original order.</li>
</ol>
<p>This results in the highest ranking observation in each experiment becoming the mean of the highest observations across all experiments, the second ranking observation in each experiment becoming the mean of the second highest observations across all experiments. Therefore the distributions for each each experiment are now the same.</p>
<p><a href="https://davetang.org/muse/2014/07/07/quantile-normalisation-in-r/">Dave Tang’s Blog:Quantile Normalisation in R</a> has more details on this approach.</p>

<div class="figure" style="text-align: center"><span id="fig:quant-norm"></span>
<img src="img/quant_norm.png" alt="Quantile Normalisation from Rafael Irizarry’s tweet." width="80%" />
<p class="caption">
Figure 5.4: Quantile Normalisation from <a href="https://twitter.com/rafalab/status/545586012219772928?ref_src=twsrc%5Etfw">Rafael Irizarry’s tweet</a>.
</p>
</div>
<p>These result of quantile normalisation is that our distributions become statisitcally identitical, which we can see by plotting the densities of the normalized data. As shown in Figure <a href="transform.html#fig:compare-normalisation">5.5</a> the distributions all overlay.</p>
<p>We do this by creating a <a href="tidyverse.html#function-anatomy">function</a>. This takes a data frame as the arguement and pefrorms the steps described to iterate through the data frame.</p>
<p>The code below is probably quite tricky to understand if you’ve not seen <code>map</code> functions before, but they enable a function such as <code>rank</code> or <code>sort</code> to be used on each column iteratively. What’s important here is to understand the aim, even if understanding the code requires some more reading. You can read about <a href="http://r4ds.had.co.nz/iteration.html#the-map-functions">map functions in R4DS</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Quantile normalisation : the aim is to give different distributions the</span>
<span class="co"># same statistical properties</span>
quantile_normalisation &lt;-<span class="st"> </span><span class="cf">function</span>(df){
  
  <span class="co"># Find rank of values in each column</span>
  df_rank &lt;-<span class="st"> </span><span class="kw">map_df</span>(df,rank,<span class="dt">ties.method=</span><span class="st">&quot;average&quot;</span>)
  <span class="co"># Sort observations in each column from lowest to highest </span>
  df_sorted &lt;-<span class="st"> </span><span class="kw">map_df</span>(df,sort)
  <span class="co"># Find row mean on sorted columns</span>
  df_mean &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(df_sorted)
  
  <span class="co"># Function for substiting mean values according to rank </span>
  index_to_mean &lt;-<span class="st"> </span><span class="cf">function</span>(my_index, my_mean){
    <span class="kw">return</span>(my_mean[my_index])
  }
  
  <span class="co"># Replace value in each column with mean according to rank </span>
  df_final &lt;-<span class="st"> </span><span class="kw">map_df</span>(df_rank,index_to_mean, <span class="dt">my_mean=</span>df_mean)
  
  <span class="kw">return</span>(df_final)
}</code></pre></div>
<p>The normalisation function is used by piping <code>dat_tidy</code> first to <code>select</code> to exclude the first two columns with the protein accession and description in, and then to the normalisation function. We re-bind the protein accession and description afterwards from <code>dat_tidy</code> by piping the output to <code>bind_cols()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_norm &lt;-<span class="st"> </span>dat_tidy <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(protein_accession<span class="op">:</span>protein_description)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">quantile_normalisation</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">bind_cols</span>(dat_tidy[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],.)</code></pre></div>

<div class="figure" style="text-align: center"><span id="fig:compare-normalisation"></span>
<img src="bspr-workshop-2018_files/figure-html/compare-normalisation-1.png" alt="Comparison of the protein distributions before normalization (left) and after quantile normalization (right)." width="80%" />
<p class="caption">
Figure 5.5: Comparison of the protein distributions before normalization (left) and after quantile normalization (right).
</p>
</div>
</div>
<div id="hypothesis-testing-with-the-t-test" class="section level2">
<h2><span class="header-section-number">5.4</span> Hypothesis testing with the t-test</h2>
<p>Having removed missing values and normalised the data, we can consider our hypothesis: treatement changes the amount of protein we observe in the cells.</p>
<p>In practice then, what we would like to know is whether the mean value for each protein in our control and treatment assays differs due to chance or due a real effect. We therefore need to calculate the difference for each protein between treatment and control, and the probability that any difference occurs due to chance. This is what the p-value from the output of a t-test seeks to do. We need to perform 1145 t-tests.</p>
<p><strong>Note</strong> There are biocondutor packages that contain functions written to do this. However as a learning exercise we are going to work through the problem.</p>
<p>Here I assume the reader is familiar with t-tests, but just to re-cap some important points:</p>
<ul>
<li><p>We assume that the true population from which our data samples are indpendent, identically distributed and follow a normal distribution. This is not in fact true in practice, but t-test is robust to this assumption.</p></li>
<li><p>We assume unequal variances between the control and treatment for each protein. Hence we will perform a Welch’s t-test for unequal variances.</p></li>
<li><p>We don’t know whether the effect of the treatment is to increase or decrease the concentration of the protein, hence we will perform a two-sided t-test.</p></li>
<li><p>The observations for the proteins are for proteins of the same type but from independent experiments, rather than observations of the same individuals before and after treatment. Hence we test the observations as unpaired samples.</p></li>
</ul>
<p>In R we use the base function <code>t.test</code> to perform Welch Two Sample t-test and this outputs the p-values we need for each protein. However, the challenge here is that our data has three observations for each condition for each protein, hence we need to group the observations for each protein according to the experimental condition as inputs to each t-test.</p>
<p>We’re going to follow what is called the <em>split-apply-combine</em> approach to deal with this problem:</p>
<ol style="list-style-type: decimal">
<li>Split the data into control and treatment groups.</li>
<li>Apply the t-test function to each protein using the grouped inputs and store the p-value.</li>
<li>Combine all the p-values for each protein into a single vector.</li>
</ol>
<p>To this end I’ve created a function called <code>t_test</code> that takes a data frame and two group vectors as inputs. It splits the data into <code>x</code> and <code>y</code> by subsetting the the data frame according to the columns defined by the groups. The extra steps here are that the subset data has to be unlisted and converted to numeric type for input to the <code>t.test</code> function. We then perform the t-test, which will calculate the mean of <code>x</code> and <code>y</code> and store the result in a new object, and finally the function creates a data frame with a single variable<code>p_val</code> which is then returned as the function output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># T-test function for multiple experiments</span>
t_test &lt;-<span class="st"> </span><span class="cf">function</span>(dt,grp1,grp2){
  <span class="co"># Subset control group and convert to numeric</span>
  x &lt;-<span class="st"> </span>dt[grp1] <span class="op">%&gt;%</span><span class="st"> </span>unlist <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.numeric</span>()
  <span class="co"># Subset treatment group and convert to numeric</span>
  y &lt;-<span class="st"> </span>dt[grp2] <span class="op">%&gt;%</span><span class="st"> </span>unlist <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.numeric</span>()
  <span class="co"># Perform t-test using the mean of x and y</span>
  result &lt;-<span class="st"> </span><span class="kw">t.test</span>(x, y)
  <span class="co"># Extract p-values from the results</span>
  p_vals &lt;-<span class="st"> </span><span class="kw">tibble</span>(<span class="dt">p_val =</span> result<span class="op">$</span>p.value)
  <span class="co"># Return p-values</span>
  <span class="kw">return</span>(p_vals)
} </code></pre></div>
<p>To use the <code>t_test</code> function to perform many t-tests and not just one t-test, we need to pass our <code>t_test</code> function as an arguement to another function.</p>
<p>This probably seems quite confusing, but the point here is that we want to loop through every row in our table, and group the three control and three treatment columns separately. Our <code>t_test</code> function deals with the latter problem, and by passing it to <code>adply</code> from the <code>plyr</code> package we can loop through each row and it adds the calculated p-values to our original table.</p>
<p>Concretely then, <code>adply</code> takes an array and applies the <code>t_test</code> function to each row and we supply the column group indices arguments to the <code>t_test</code> funcition. Here the indicies are columns 3 to 5 for the control experiments and columns 6 to 8 for the treatment functions. The function returns the input data with an additional corresponding p-value column. <strong>Note</strong> I’ve piped the output to <code>as.tibble()</code> to transform the data.frame output of <code>adply</code> to tibble form to prevent errors that can occur if we try to bind data frames and tibbles.</p>
<p>An important point here is that we can use this function for any number of columns and rows providing our data is in the same tidy form by changing the grouping indices.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Apply t-test function to data using plyr adply</span>
<span class="co">#  .margins = 1, slice by rows, .fun = t_test plus t_test arguements</span>
dat_pvals &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">adply</span>(dat_norm,<span class="dt">.margins =</span> <span class="dv">1</span>, <span class="dt">.fun =</span> t_test, 
                <span class="dt">grp1 =</span> <span class="kw">c</span>(<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>), <span class="dt">grp2 =</span> <span class="kw">c</span>(<span class="dv">6</span><span class="op">:</span><span class="dv">8</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.tibble</span>()</code></pre></div>
<p>To check our function, here’s a comparision of calculating the first protein p-value as a single t-test as shown in the following code and the output of the function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Perform t-test on first protein</span>
<span class="kw">t.test</span>(<span class="kw">as.numeric</span>(dat_norm[<span class="dv">1</span>,<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>]),
                    <span class="kw">as.numeric</span>(dat_norm[<span class="dv">1</span>,<span class="dv">6</span><span class="op">:</span><span class="dv">8</span>]))<span class="op">$</span>p.value</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="right">t_test p-val</th>
<th align="right">t.test p-val</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">0.0927</td>
<td align="right">0.0927</td>
</tr>
</tbody>
</table>
<p>We can plot a histogram of the p-values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot histogram</span>
dat_pvals <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(p_val)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">binwidth =</span> <span class="fl">0.05</span>, 
           <span class="dt">boundary =</span> <span class="fl">0.5</span>, 
           <span class="dt">fill =</span> <span class="st">&quot;darkblue&quot;</span>,
           <span class="dt">colour =</span> <span class="st">&quot;white&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;p-value&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Frequency&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_minimal</span>()</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/p-val-plot-1.png" width="672" /></p>
</div>
<div id="calculating-fold-change" class="section level2">
<h2><span class="header-section-number">5.5</span> Calculating fold change</h2>
<p>To perform log transformation of the observations for each protein we take our data and use select to exlude the columns of character vectors and the pipe the output to <code>log2()</code> and use the pipe again to create a data frame.</p>
<p>Then we use <code>bind_cols</code> to bind the first two columns of <code>dat_pvals</code> followed by <code>dat_log</code> and the last column of <code>dat_pvals</code>. This maintains the original column order.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Select columns and log data</span>
dat_log &lt;-<span class="st"> </span>dat_pvals <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">c</span>(protein_accession,protein_description,p_val)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">log2</span>()

<span class="co"># Bind columns to create transformed data frame</span>
dat_combine &lt;-<span class="st"> </span><span class="kw">bind_cols</span>(dat_pvals[,<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)], dat_log, dat_pvals[,<span class="dv">9</span>]) </code></pre></div>
<p>The log fold change is then the difference between the log mean control and log mean treatment values. By use of grouping by the protein accession we can then use <code>mutate</code> to create new variables that calculate the mean values and then calculate the <code>log_fc</code>. Whilst we’re about it, we can also calculate a -log10(p-value). As with fold change, transforming the p-value on a log10 scale means that a p-value of 0.05 or below is transformed to 1.3 or above and a p-value of 0.01 is equal to 2.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_fc &lt;-<span class="st"> </span>dat_combine <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(protein_accession) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">mean_control =</span> <span class="kw">mean</span>(<span class="kw">c</span>(control_<span class="dv">1</span>,
                               control_<span class="dv">2</span>,
                               control_<span class="dv">3</span>)),
                             <span class="dt">mean_treatment=</span> <span class="kw">mean</span>(<span class="kw">c</span>(treatment_<span class="dv">1</span>,
                                                    treatment_<span class="dv">2</span>,
                                                    treatment_<span class="dv">3</span>)),
         <span class="dt">log_fc =</span> mean_control <span class="op">-</span><span class="st"> </span>mean_treatment,
         <span class="dt">log_pval =</span> <span class="op">-</span><span class="dv">1</span><span class="op">*</span><span class="kw">log10</span>(p_val))</code></pre></div>
<p>The next step is not necessary, but for ease of viewing we subset <code>dat_fc</code> to create a new data frame called <code>dat_tf</code> that contains only four variables. We could potentially write this to a csv file for sharing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Final transformed data</span>
dat_tf &lt;-<span class="st"> </span>dat_fc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(protein_accession,
                            protein_description,
                            log_fc, log_pval)</code></pre></div>
<p>Let’s look at the head of the final table:</p>
<table>
<thead>
<tr class="header">
<th align="left">protein_accession</th>
<th align="left">protein_description</th>
<th align="right">log_fc</th>
<th align="right">log_pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">VATA_HUMAN_P38606</td>
<td align="left">V-type proton ATPase catalytic subunit A OS=Homo sapiens GN=ATP6V1A PE=1 SV=2</td>
<td align="right">0.3687886</td>
<td align="right">1.0327506</td>
</tr>
<tr class="even">
<td align="left">RL35A_HUMAN_P18077</td>
<td align="left">60S ribosomal protein L35a OS=Homo sapiens GN=RPL35A PE=1 SV=2</td>
<td align="right">-0.2505780</td>
<td align="right">1.1400318</td>
</tr>
<tr class="odd">
<td align="left">MYH10_HUMAN_P35580</td>
<td align="left">Myosin-10 OS=Homo sapiens GN=MYH10 PE=1 SV=3</td>
<td align="right">0.3838733</td>
<td align="right">0.0056494</td>
</tr>
<tr class="even">
<td align="left">RHOG_HUMAN_P84095</td>
<td align="left">Rho-related GTP-binding protein RhoG OS=Homo sapiens GN=RHOG PE=1 SV=1</td>
<td align="right">-0.3417452</td>
<td align="right">0.3775483</td>
</tr>
<tr class="odd">
<td align="left">PSA1_HUMAN_P25786</td>
<td align="left">Proteasome subunit alpha type-1 OS=Homo sapiens GN=PSMA1 PE=1 SV=1</td>
<td align="right">0.0371316</td>
<td align="right">0.0920101</td>
</tr>
</tbody>
</table>
</div>
<div id="visualising-the-transformed-data" class="section level2">
<h2><span class="header-section-number">5.6</span> Visualising the transformed data</h2>
<p>Plotting a histogram of the log fold change gives an indication of whether the treatment has an effect on the cells. Most values are close to zero, but there are some observations far above and below zero suggesting the treatment does have an effect.</p>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Plot a histogram to look at the distribution.</span>
dat_tf <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(log_fc)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">binwidth =</span> <span class="fl">0.5</span>,
                 <span class="dt">boundary =</span> <span class="fl">0.5</span>,
           <span class="dt">fill =</span> <span class="st">&quot;darkblue&quot;</span>,
           <span class="dt">colour =</span> <span class="st">&quot;white&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;log2 fold change&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Frequency&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_minimal</span>()</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:log-fc"></span>
<img src="bspr-workshop-2018_files/figure-html/log-fc-1.png" alt="Histogram of log fold change." width="80%" />
<p class="caption">
Figure 5.6: Histogram of log fold change.
</p>
</div>
<p>However, we don’t know if these fold changes are dueto chance or not, which is why we calculated the p-values. A volcano plot will include the p-value information.</p>
</div>
<div id="volcano-plot" class="section level2">
<h2><span class="header-section-number">5.7</span> Volcano plot</h2>
<!--- 7. Create a combined table of log fold change and p-values for all the proteins
for plotting a volcano plot. --->
<p>A volcano plot is a plot of the log fold change in the observation between two conditions on the x-axis, for example the protein expression between treatment and control conditions. On the y-axis is the corresponding p-value for each observation, representing the likelihood that an observed change is due to the different conditions rather than arising from a natural variation in the fold change that might be observed if we performed many replications of the experiment.</p>
<p>The aim of a volcano plot is to enable the viewer to quickly see the effect (if any) of an experiment with two conditions on many species (i.e. proteins) in terms of both an increase and decrease of the observed value.</p>
<p>Like all plots it has it’s good and bad points, namely it’s good that we can visualise a lot of complex information in one plot. However this is also it’s main weakness, it’s rather complicated to understand in one glance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_tf <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(log_fc,log_pval)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="bspr-workshop-2018_files/figure-html/volcano-plot-1.png" width="672" /></p>
<p>However it would be much more useful with some extra formatting, so the code below shows one way to transform the data to include a threshold which can then be used by ggplot to create an additional aesthetic. The code below also includes some extra formatiing which the reader can explore.</p>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_tf <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Add a threhold for significant observations</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">threshold =</span> <span class="kw">if_else</span>(log_fc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>log_pval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.3</span> <span class="op">|</span>
<span class="st">                               </span>log_fc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">2</span> <span class="op">&amp;</span><span class="st"> </span>log_pval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.3</span>,<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Plot with points coloured according to the threshold</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(log_fc,log_pval, <span class="dt">colour =</span> threshold)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span><span class="co"># Alpha sets the transparency of the points</span>
<span class="st">  </span><span class="co"># Add dotted lines to indicate the threshold, semi-transparent</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">1.3</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">2</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="op">-</span><span class="dv">2</span>, <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set the colour of the points</span>
<span class="st">  </span><span class="kw">scale_colour_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>=<span class="st"> &quot;red&quot;</span>, <span class="st">&quot;B&quot;</span>=<span class="st"> &quot;black&quot;</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;log2 fold change&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;-log10 p-value&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="co"># Relabel the axes</span>
<span class="st">  </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span><span class="co"># Set the theme</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>) <span class="co"># Hide the legend</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:nice-vplot"></span>
<img src="bspr-workshop-2018_files/figure-html/nice-vplot-1.png" alt="A volcano plot with formatting to highlight the significant proteins" width="80%" />
<p class="caption">
Figure 5.7: A volcano plot with formatting to highlight the significant proteins
</p>
</div>
<div id="but-which-proteins-are-the-significant-observations" class="section level3">
<h3><span class="header-section-number">5.7.1</span> But which proteins are the significant observations?</h3>
<p>To extract the proteins in red in Figure <a href="transform.html#fig:nice-vplot">5.7</a> we filter <code>dat_tf</code> according to our threshold and then create a new variable using the <code>str_extract</code> function used in Section <a href="dplyr.html#mutate">4.5</a>.</p>
<p><strong>Note</strong> We need to ungroup the data we grouped when calculating the log_fc to be able to select columns without keeping the grouping variable column too.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_tf <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># Filter for significant observations</span>
<span class="st">  </span><span class="kw">filter</span>(log_pval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.3</span> <span class="op">&amp;</span><span class="st"> </span>(log_fc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2</span> <span class="op">|</span><span class="st"> </span>log_fc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">2</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Get last six characters</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">prot_id =</span> <span class="kw">str_extract</span>(protein_accession,<span class="st">&quot;.{6}$&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Ungroup the data</span>
<span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Select columns of interest</span>
<span class="st">  </span><span class="kw">select</span>(prot_id,protein_description,log_fc,log_pval)</code></pre></div>
<pre><code>## # A tibble: 5 x 4
##   prot_id protein_description                              log_fc log_pval
##   &lt;chr&gt;   &lt;chr&gt;                                             &lt;dbl&gt;    &lt;dbl&gt;
## 1 Q02952  A-kinase anchor protein 12 OS=Homo sapiens GN=A~  -2.29     1.52
## 2 O94808  Glutamine--fructose-6-phosphate aminotransferas~  -3.09     2.45
## 3 H7BYV1  Interferon-induced transmembrane protein 2 (Fra~   2.05     2.46
## 4 P06756  Integrin alpha-V OS=Homo sapiens GN=ITGAV PE=1 ~  -2.22     2.62
## 5 Q8TDI0  Chromodomain-helicase-DNA-binding protein 5 OS=~   5.04     1.60</code></pre>
</div>
</div>
<div id="creating-a-heatmap" class="section level2">
<h2><span class="header-section-number">5.8</span> Creating a heatmap</h2>
<p>Here we’ll create a heatmap using the <code>heatmap.2</code> function from the <code>gplots</code> package and the <code>pheatmap</code> function from the <code>pheatmap</code> package.</p>
<p>To create a heatmap we need to perform a few more transformations:</p>
<ol style="list-style-type: decimal">
<li><p>Filter the data according to a threshold of significance. This time we’ll use a more relaxed log_fc cut-off to ensure we have enough proteins to plot. At the same time we’ll extract the protein ids as before.</p></li>
<li><p>We then have to transform our filtered data into a <code>matrix.data.frame</code> object for use with <code>pheatmap</code>. We name the rows with the protein ids</p></li>
<li><p>We’ll use base R function <code>scale</code> to centre our log transformed data around zero. To do this per experiment we transpose the matrix as scale centres rows, and the flip the matrix back again.</p></li>
</ol>
<!-- 1. Clustering the data -->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Keep the same p-val cut-off, but relax the log_fc to 1 which represents a </span>
<span class="co"># doubling</span>
dat_filt &lt;-<span class="st"> </span>dat_fc <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(log_pval <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.3</span> <span class="op">&amp;</span><span class="st"> </span>(log_fc <span class="op">&gt;=</span><span class="st"> </span><span class="dv">1</span> <span class="op">|</span><span class="st"> </span>log_fc <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="dv">1</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">prot_id =</span> <span class="kw">str_extract</span>(protein_accession,<span class="st">&quot;.{6}$&quot;</span>))

<span class="co"># Convert to matrix data frame</span>
dat_matrix &lt;-<span class="st"> </span><span class="kw">as.matrix.data.frame</span>(dat_filt[,<span class="dv">3</span><span class="op">:</span><span class="dv">8</span>]) 
<span class="co"># Name the rows with protein ids</span>
<span class="kw">row.names</span>(dat_matrix) &lt;-<span class="st"> </span>dat_filt<span class="op">$</span>prot_id
<span class="co"># Transpose and scale the data to a mean of zero and sd of one</span>
dat_scaled &lt;-<span class="st"> </span><span class="kw">scale</span>(<span class="kw">t</span>(dat_matrix)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">t</span>()</code></pre></div>
<div id="calculating-similarity-and-clustering" class="section level3">
<h3><span class="header-section-number">5.8.1</span> Calculating similarity and clustering</h3>
<p>At this point we could just plot the data, but to understand what the heatmap functions do to cluster the data, let’s step through the process.</p>
<p>Our data here as log fold change in concentrations, but how do we group them? The simplest thing to do is to turn the data into distances, as a measure of similarity, where close things are similar and distant things are dissimilar.</p>
<p>The Euclidean distance <span class="math inline">\(d\)</span> between a pair of observations <span class="math inline">\(x_i\)</span> and <span class="math inline">\(y_i\)</span> is defined as:</p>
<p><span class="math inline">\(d = \sqrt{\sum{_i}(x_i - y_i)^2}\)</span></p>
<p>Lets calculate the distance between the columns in <code>dat_scaled</code>.</p>
<p>In <code>dat_scaled</code> the experiments are in the columns. In calculating the distance is between the experiments for all the proteins in each experiment. What would we expect?</p>
<p>We’d expect the controls to be close to each other and the treated to be close to each other, right?</p>
<p>Let’s do this in detail, for example the distance between <code>control_1</code> and <code>control_2</code> is <code>sqrt(sum((dat_scaled[,1] - dat_scaled[,2])^2))</code>.</p>
<p>This means we take the column 2 values from column 1 values, squaring the results and summing them all to a single value and taking the square root to find the linear distance between these rows, which is 3.26.</p>
<p>You can check this against the first value in <code>d1</code> that we calculate below in using <code>dist</code>.</p>
<p>We do the same for the proteins, but we don’t know what to expect. Here’s the code for calculating both distance matrices</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Transpose the matrix to calculate distance between experiments, row-wise</span>
d1 &lt;-<span class="st"> </span>dat_scaled <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">t</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">dist</span>(.,<span class="dt">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">diag =</span> <span class="ot">FALSE</span>, <span class="dt">upper =</span> <span class="ot">FALSE</span>)
<span class="co"># Calculate the distance between proteins row-wise </span>
d2 &lt;-<span class="st"> </span>dat_scaled <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">dist</span>(.,<span class="dt">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">diag =</span> <span class="ot">FALSE</span>, <span class="dt">upper =</span> <span class="ot">FALSE</span>)

<span class="co"># Show the values for d1</span>
<span class="kw">round</span>(d1,<span class="dv">2</span>)</code></pre></div>
<pre><code>##             control_1 control_2 control_3 treatment_1 treatment_2
## control_2        3.26                                            
## control_3        3.20      3.27                                  
## treatment_1      8.97      8.60      8.65                        
## treatment_2      9.40      8.98      8.86        2.35            
## treatment_3      9.04      8.56      8.50        2.46        1.71</code></pre>
<p>Having calculated the distance matrices, we can cluster proteins and experiments accordingly.</p>
<p>There are lots of flavours of clustering, and no clear way to say which is best. Here we’ll use the Ward criterion for clustering which attempts to minimise the variance within clusters as it merges the data into clusters, using the distances we’ve calculated. The data is merged from the bottom up (aka agglomeration) adding data points to a cluster and splitting them according to the variance criterion.</p>
<p>See Wikipedia for more detail: <a href="https://en.wikipedia.org/wiki/Hierarchical_clustering">Hierarchical clustering</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Clustering distance between experiments using Ward linkage</span>
c1 &lt;-<span class="st"> </span><span class="kw">hclust</span>(d1, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="dt">members =</span> <span class="ot">NULL</span>)
<span class="co"># Clustering distance between proteins using Ward linkage</span>
c2 &lt;-<span class="st"> </span><span class="kw">hclust</span>(d2, <span class="dt">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="dt">members =</span> <span class="ot">NULL</span>)</code></pre></div>
<p>Now lets look at the dendrograms made by clustering our distance matrices <code>d1</code> and <code>d2</code>:</p>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Check clustering by plotting dendrograms</span>
<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">cex=</span><span class="fl">0.5</span>) <span class="co"># Make 2 rows, 1 col plot frame and shrink labels</span>
<span class="kw">plot</span>(c1); <span class="kw">plot</span>(c2) <span class="co"># Plot both cluster dendrograms</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:dendrograms"></span>
<img src="bspr-workshop-2018_files/figure-html/dendrograms-1.png" alt="Dendrograms of Ward clustering of distance matrices" width="80%" />
<p class="caption">
Figure 5.8: Dendrograms of Ward clustering of distance matrices
</p>
</div>
<p>As we’d expect, Figure <a href="transform.html#fig:dendrograms">5.8</a> shows the controls and treatments cluster respectively.</p>
<!-- 2. Plot the data -->
</div>
<div id="plotting-the-heatmap" class="section level3">
<h3><span class="header-section-number">5.8.2</span> Plotting the heatmap</h3>
<p>The <code>heatmap.2</code> function from the <code>gplots</code> package will automatically perform the distance calculation and clustering we performed, and it can also do the scaling we did. It only requires the matrix as an input by default. It will use a different clustering method by default.</p>
<p>However, as we’ve performed scaling and calculated the clusters, we can pass them to heatmap function.</p>
<p>I’ll leave it to the reader to explore all the options here, but the concept in the code below to create Figure <a href="transform.html#fig:heatmap2">5.9</a> is:</p>
<ul>
<li>Create a 25 increment blue/white/red colour pallette</li>
<li>Pipe <code>dat_scaled</code> to a function that renames the colums</li>
<li>Pipe this to the <code>heatmap.2</code> function</li>
<li>Pass the clusters <code>c1</code> and <code>c2</code> to the plot</li>
<li>Change some aesthetics such as the colours, and the font sizes</li>
</ul>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Set colours for heatmap, 25 increments</span>
my_palette &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;white&quot;</span>,<span class="st">&quot;red&quot;</span>))(<span class="dt">n =</span> <span class="dv">25</span>)

<span class="co"># Plot heatmap with heatmap.2</span>
<span class="kw">par</span>(<span class="dt">cex.main=</span><span class="fl">0.75</span>) <span class="co"># Shrink title fonts on plot</span>
dat_scaled <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Rename the comlums</span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_colnames</span>(<span class="kw">c</span>(<span class="st">&quot;Ctl 1&quot;</span>, <span class="st">&quot;Ctl 2&quot;</span>, <span class="st">&quot;Ctl 3&quot;</span>,
                                    <span class="st">&quot;Trt 1&quot;</span>, <span class="st">&quot;Trt 2&quot;</span>, <span class="st">&quot;Trt 3&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Plot heatmap</span>
<span class="st">  </span>gplots<span class="op">::</span><span class="kw">heatmap.2</span>(.,                     <span class="co"># Tidy, normalised data</span>
          <span class="dt">Colv=</span><span class="kw">as.dendrogram</span>(c1),     <span class="co"># Experiments clusters in cols</span>
          <span class="dt">Rowv=</span><span class="kw">as.dendrogram</span>(c2),     <span class="co"># Protein clusters in rows</span>
          <span class="dt">revC=</span><span class="ot">TRUE</span>,                  <span class="co"># Flip plot to match pheatmap</span>
          <span class="dt">density.info=</span><span class="st">&quot;histogram&quot;</span>,   <span class="co"># Plot histogram of data and colour key</span>
          <span class="dt">trace=</span><span class="st">&quot;none&quot;</span>,               <span class="co"># Turn of trace lines from heat map</span>
          <span class="dt">col =</span> my_palette,           <span class="co"># Use my colour scheme</span>
          <span class="dt">cexRow=</span><span class="fl">0.6</span>,<span class="dt">cexCol=</span><span class="fl">0.75</span>)     <span class="co"># Amend row and column label fonts</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:heatmap2"></span>
<img src="bspr-workshop-2018_files/figure-html/heatmap2-1.png" alt="Heatmap created with heatmap.2 using the clusters calculated." width="80%" />
<p class="caption">
Figure 5.9: Heatmap created with <code>heatmap.2</code> using the clusters calculated.
</p>
</div>
<p>An alternative and more <code>ggplot</code> style is to use the <code>pheatmap</code> package and function <span class="citation">(Kolde <a href="#ref-R-pheatmap">2018</a>)</span>.</p>
<p>In Figure <a href="transform.html#fig:pheatmap">5.10</a> <code>dat_scaled</code> is piped to <code>set_columns</code> again to rename the experiments for aesthetic reasons. The output is the piped to <code>pheatmap</code> which performs the distance and clustering automatically. The only additional arguements used here are to change the fontsize and create some breaks in the plot to highlight the clustering.</p>
<p>There is lots more that <code>pheatmap</code> can do in terms of aesthetics, so do explore.</p>

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat_scaled <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Rename the comlums</span>
<span class="st">  </span>magrittr<span class="op">::</span><span class="kw">set_colnames</span>(<span class="kw">c</span>(<span class="st">&quot;Ctl 1&quot;</span>, <span class="st">&quot;Ctl 2&quot;</span>, <span class="st">&quot;Ctl 3&quot;</span>,
                                    <span class="st">&quot;Trt 1&quot;</span>, <span class="st">&quot;Trt 2&quot;</span>, <span class="st">&quot;Trt 3&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Plot heatmap</span>
<span class="st">  </span><span class="kw">pheatmap</span>(.,
           <span class="dt">fontsize =</span> <span class="dv">7</span>,
           <span class="dt">cutree_rows =</span> <span class="dv">2</span>, <span class="co"># Create breaks in heatmap</span>
           <span class="dt">cutree_cols =</span> <span class="dv">2</span>) <span class="co"># Create breaks in heatmap</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:pheatmap"></span>
<img src="bspr-workshop-2018_files/figure-html/pheatmap-1.png" alt="Heatmap created using pheatmap with breaks to highlight clusters." width="80%" />
<p class="caption">
Figure 5.10: Heatmap created using <code>pheatmap</code> with breaks to highlight clusters.
</p>
</div>
</div>
</div>
<div id="venn-diagram" class="section level2">
<h2><span class="header-section-number">5.9</span> Venn diagram</h2>
<p>Another common plot used in proteomics is the Venn diagram. For these I use the <code>VennDiagram</code> package <span class="citation">(Chen <a href="#ref-R-VennDiagram">2018</a>)</span></p>
</div>
<div id="peptide-sequence-logos" class="section level2">
<h2><span class="header-section-number">5.10</span> Peptide sequence logos</h2>
<p>Finally, creating sequence logos from peptides is another common task, especially if you are doing immunopeptidomics and would like to explore the fequency of amino acid types at each position in a set of peptide sequences.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-bolstad2003">
<p>Bolstad, B M, R A Irizarry, M Astrand, and T P Speed. 2003. “A Comparison of Normalization Methods for High Density Oligonucleotide Array Data Based on Variance and Bias.” <em>Bioinformatics (Oxford, England)</em> 19 (2): 185–93.</p>
</div>
<div id="ref-R-pheatmap">
<p>Kolde, Raivo. 2018. <em>Pheatmap: Pretty Heatmaps</em>. <a href="https://CRAN.R-project.org/package=pheatmap" class="uri">https://CRAN.R-project.org/package=pheatmap</a>.</p>
</div>
<div id="ref-R-VennDiagram">
<p>Chen, Hanbo. 2018. <em>VennDiagram: Generate High-Resolution Venn and Euler Plots</em>. <a href="https://CRAN.R-project.org/package=VennDiagram" class="uri">https://CRAN.R-project.org/package=VennDiagram</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dplyr.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="going-further.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["bspr-workshop-2018.pdf", "bspr-workshop-2018.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
